keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+    // was allocated. They can be arbitrary setup parameters that are
+    // ignore by the counter controller.
+    VendorClassParameters *runtime.RawExtension
+
+    // VendorClaimParameters are the parameters from the time that the claim was
+    // allocated. Some of the fields were used by the counter controller to
+    // allocated resources, others can be arbitrary setup parameters.
+    VendorClaimParameters *runtime.RawExtension
+
+   	// NodeName is the name of the node providing the necessary resources.
+    // This mirrors the AllocationResult.AvailableOnNodes with a simpler
+    // type.
+    //
+    // The driver name is the one stored in ResourceHandle.
+    NodeName string
+
+    AllocationResultModel // inline, fields must not conflict
+}
+
+// AllocationResultModel must have one and only one field set.
+type AllocationResultModel struct {
+    Partitionable *PartitionableAllocationResult
+}
+
+type PartitionableAllocationResult struct {
+    Instances []AllocatedResourceInstance
+}
+
+type AllocatedResourceInstance struct {
+   ID string // A concatenation with / of the individual names.
+}
+```
+
+### Implementation of semantic models
+
+In the Go types above, all structs starting with `Partitionable` are part of that
+semantic model. In practice, organizing those inside their own Go package and
+then importing that package in the definition of the resource.k8s.io API will
+result in a cleaner separation at the source code level. It has no impact on
+the resulting Kubernetes API.
+
+### Scheduling + Allocation
+
+The dynamic resource scheduler plugin handles the common fields of
+NodeResourceSlice, ResourceClaimParameters and SemanticResourceHandle. For the
+semantic model fields it calls out to code that is associated with the
+corresponding model.
+
+During filtering it is decided which nodes have the necessary resources. If a
+node is found, the scheduler plugin updates the resource claim status as part
+of goroutine which handles pod binding.
+
+Like a normal DRA driver controller, the scheduler also sets a finalizer to
+ensure that users cannnot accidentally delete the allocated claim while a pod
+is about to start which depends on it. That finalizer is
+"semantic.dra.k8s.io/delete-protection".
+
+### Deallocation
+
+Deallocation is handled by kube-controller-manager when its claim controller
+observes that a claim is no longer in use *and* the claim has the special
+"semantic.dra.k8s.io/delete-protection" finalizer. This finalizer tells the
+controller that it may clear the allocation result directly instead of setting
+the `DeletionRequested` field, which is what it normally would do.
+
+Updating the claim during deallocation will be observed by kube-scheduler and
+tells it that it can use the capacity set aside for the claim
+again. kube-controller-manager itself doesn't need to support specific semantic
+models.
+
+### Immediate allocation
+
+Because there is no separate controller anymore, claims with immediate
```

Comment:
We need to look at this anew before beta.  Is this still representing user intention well?

Reply:
> claims with immediate allocation remain allocated when no longer in use

I think this is the important bit -- what use cases might require this (and why)?

I don't have any "immediately" on hand (pun intended!).

Reply:
AIUI, part of the motivation for immediate allocation was for expensive initialization. Suppose you need to flash a card, which takes a long time, then want to run a whole bunch of workloads against it in series (rather than in parallel). @pohly ?

Reply:
Even in the "existing" DRA model, no configuration happens until the first pod that uses a claim lands on the node where it has been allocated. That first pod will always be delayed by the time to configure in both scenarios.

The real difference with this KEP is that now the allocation itself won't happen until a pod that references the claim is scheduled. Whereas before, the DRA controller would pick an arbitrary node to allocate the claim on as soon as the claim object itself was created, and the scheduler was forced to schedule any pods referencing that claim on that node.

Reply:
Are we saying that "allocation mode" is defunct?

Reply:
I think this is a different topic. I had suggested adding an optional nodeName to resource claim in an earlier model. I cannot abide the "randomly pick a node". The nodeName would allow immediate allocation without that, and allow then for pods to be attached to that same (shared) claim.

But I don't think that's key in this KEP, I think that's a discussion for how we evolve base DRA in 1.31.

Reply:
I agree we should hold off on this, but i would also like to express my disdain for "randomly pick a node". It's the main reason we never built support for this mode in the existing DRA driver for GPUs.